openedx_external_enrollments/external_enrollments.py

```diff
@@ -0,0 +1,424 @@
+"""External enrollments file."""
+import datetime
+import logging
+from collections import OrderedDict
+
+import requests
+from django.conf import settings
+from oauthlib.oauth2 import BackendApplicationClient
+from opaque_keys.edx.keys import CourseKey
+from requests_oauthlib import OAuth2Session
+from rest_framework import status
+
+from openedx_external_enrollments.edxapp_wrapper.get_courseware import get_course_by_id
+from openedx_external_enrollments.edxapp_wrapper.get_student import CourseEnrollment, get_user
+from openedx_external_enrollments.models import (
+    EnrollmentRequestLog,
+    ProgramSalesforceEnrollment,
+)
+
+LOG = logging.getLogger(__name__)
+
+
+class BaseExternalEnrollment(object):
+    """
+    """
+    def _execute_post(self, url, data=None, headers=None, json_data=None):
+        """
+        """
+        response = requests.post(
+            url=url,
+            data=data,
+            headers=headers,
+            json=json_data,
+        )
+        return response
+
+    def _post_enrollment(self, data, course_settings=None):
+        """
+        """
+        url = self._get_enrollment_url(course_settings)
+        json_data = self._get_enrollment_data(data, course_settings)
+        LOG.info('calling enrollment for [%s] with data: %s', self.__str__(), json_data)
+        LOG.info('calling enrollment for [%s] with url: %s', self.__str__(), url)
+        LOG.info('calling enrollment for [%s] with course settings: %s', self.__str__(), course_settings)
+        log_details = {
+            "request_payload": json_data,
+            "url": url,
+            "course_advanced_settings": course_settings,
+        }
+
+        try:
+            response = self._execute_post(
+                url=url,
+                headers=self._get_enrollment_headers(),
+                json_data=json_data,
+            )
+        except Exception as error:
+            LOG.error("Failed to complete enrollment. Reason: " + str(error))
+            log_details["response"] = {"error": "Failed to complete enrollment. Reason: " + str(error)}
+            EnrollmentRequestLog.objects.create(request_type=str(self), details=log_details)
```

Comment:
What should happen if .create fails? @andrey-canon

Reply:
out of the scope of this pr